### PR TITLE
ASoC/SoundWire: add support for RT712VB quirk

### PR DESCRIPTION
**Pull request by plbossart**
Simpler version that what was shared before in https://github.com/thesofproject/linux/pull/4995

No major change other than better data structures and preparation for more Function-level support.

For now we do need the RT712VB quirk so that PR adds only this "small" part which is already rather intrusive between SoundWire and ASoC/SOF domains.

### CODE REVIEW

File: sound/soc/intel/common/soc-acpi-intel-mtl-match.c [hunk 1]
```diff
@@ -715,6 +756,31 @@ static const struct snd_soc_acpi_link_adr cs42l43_link0_cs35l56_link2_link3[] =
...
 
+/*
+ * Pretend machine quirk. The argument type is not the traditional
+ * 'struct snd_soc_acpi_mach' pointer but instead the sdw_intel_ctx
+ * which contains the peripheral information required for the
+ * SoundWire/SDCA filter on the SMART_MIC setup and interface
+ * revision. When the return value is false, the entry in the
+ * 'snd_soc_acpi_mach' table needs to be skipped.
+ */
+static bool is_device_rt712_vb(void *arg)
```
Comment (bardliao) [inline sound/soc/intel/common/soc-acpi-intel-mtl-match.c:767]: I am thinking will it be better to add `is_device_rt712_vb` in a generic file? LNL+ platforms will need this, too.

Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-mtl-match.c:767]: good suggestion @bardliao, will move the code.

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -899,8 +899,15 @@ static struct sof_sdw_codec_info codec_info_list[] = {
...
+			{
+				.direction = {false, true},
+				.dai_name = "rt712-sdca-aif3",
+				.dai_type = SOF_SDW_DAI_TYPE_MIC,
+				.dailink = {SDW_UNUSED_DAI_ID, SDW_DMIC_DAI_ID},
+				.rtd_init = rt_dmic_rtd_init,
+			},
 		},
-		.dai_num = 2,
+		.dai_num = 3,
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_sdw.c:910]: rt713 has -vb version, too.
```
diff --git a/sound/soc/intel/boards/sof_sdw.c b/sound/soc/intel/boards/sof_sdw.c
index 87e90e9a90a8..8fd02af61dd4 100644
--- a/sound/soc/intel/boards/sof_sdw.c
+++ b/sound/soc/intel/boards/sof_sdw.c
@@ -951,8 +951,15 @@ static struct sof_sdw_codec_info codec_info_list[] = {
                                .widgets = generic_jack_widgets,
                                .num_widgets = ARRAY_SIZE(generic_jack_widgets),
                        },
+                       {
+                               .direction = {false, true},
+                               .dai_name = "rt712-sdca-aif3",
+                               .dai_type = SOF_SDW_DAI_TYPE_MIC,
+                               .dailink = {SDW_UNUSED_DAI_ID, SDW_DMIC_DAI_ID},
+                               .rtd_init = rt_dmic_rtd_init,
+                       },
                },
-               .dai_num = 1,
+               .dai_num = 2,
        },
        {
                .part_id = 0x1713,
```

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:910]: that's what I asked earlier, I don't understand which part does what :-)

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:910]: > that's what I asked earlier, I don't understand which part does what :-)

The -va version uses a separated sdw interface (rt1712/rt1713) for DMIC, and -vb uses the same sdw interface. -vb = -va + dmic, rt1713-va/rt1712-va = dmic.
And, rt713 = rt712 - amp,  
So, rt712-va = jack + amp, rt713-va = jack, rt712-vb = jack + amp + dmic, rt713-vb = jack + dmic, rt1713-va/rt1712-va = dmic

**Comment by plbossart** [comment]
closing, will re-add a new branch

**Comment by plbossart** [comment]
closing, will re-add a new branch

**Pull request closed without merging**